pkg/plugins/common/postgres/pgx_listener.go

```diff
@@ -0,0 +1,209 @@
+package postgres
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/jackc/pgx/v5/pgconn"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"github.com/pkg/errors"
+
+	"github.com/kumahq/kuma/pkg/config/plugins/resources/postgres"
+)
+
+// PgxListener will listen for NOTIFY commands on a channel.
+type PgxListener struct {
+	notificationsCh chan *Notification
+
+	logger logr.Logger
+
+	channel string
+	conn    *pgxpool.Conn
+	ctx     context.Context
+	db      *pgxpool.Pool
+
+	stopFn    func()
+	stoppedCh chan struct{}
+
+	mx      sync.Mutex
+	running bool
+}
+
+var _ Listener = (*PgxListener)(nil)
+
+// NewPgxListener will create and initialize a PgxListener which will automatically reconnect and listen to the provided channels.
+func NewPgxListener(config postgres.PostgresStoreConfig, logger logr.Logger) (Listener, error) {
+	ctx := context.Background()
+	connectionString, err := config.ConnectionString()
+	if err != nil {
+		return nil, err
+	}
+	db, err := pgxpool.New(context.Background(), connectionString)
+	if err != nil {
+		return nil, err
+	}
+	l := &PgxListener{
+		notificationsCh: make(chan *Notification, 32),
+		logger:          logger,
+		ctx:             ctx,
+		channel:         "resource_events",
+		db:              db,
+	}
+
+	err = l.connect(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	l.Start()
+
+	return l, nil
+}
+
+func (l *PgxListener) Start() {
+	l.logger.Info("started")
+
+	l.mx.Lock()
+	defer l.mx.Unlock()
+
+	if l.running {
+		return
+	}
+
+	ctx, cancel := context.WithCancel(l.ctx)
+	l.stopFn = cancel
+	l.stoppedCh = make(chan struct{}, 1)
+
+	go l.run(ctx)
+
+	l.running = true
+}
+
+func (l *PgxListener) Close() error {
+	l.stop()
+	close(l.notificationsCh)
+	return nil
+}
+
+func (l *PgxListener) run(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			l.stoppedCh <- struct{}{}
+			return
+		default:
+		}
+		err := l.handleNotifications(ctx)
+		if errors.Is(err, context.Canceled) {
+			err = nil
+		}
+		if err != nil {
+			l.logger.Error(err, "error handling notification")
+		}
+	}
+}
+
+// stop will end all current connections and stop reconnecting.
+func (l *PgxListener) stop() {
+	l.logger.Info("stopped")
```

Comment:
technically it's stopping, but not yet stopped

Reply:
moved